File: compiler/src/phase5/persistent-learning-loop.ts
// ============================================================================
// phase5/persistent-learning-loop.ts
// Phase 5 - 持久化学习循环
//
// 在 learning-loop.ts 的基础上增加：
//   1. JSON文件持久化存储（替代MemoryCaseStore的内存存储）
//   2. 自动保存/加载机制（每次操作后自动保存）
//   3. 检查点机制（定期保存检查点）
//   4. 跨IDE共享（文件系统存储）
//
// 对应架构设计文档 7.5 节 LearningLoop + 持久化扩展
// ============================================================================

import * as fs from "fs";
import * as path from "path";
import {
    ExecutionRecord,
    ParameterTemplate,
    ConfidenceAdjustment,
    LearningMetrics,
    VerificationResult,
    ExpectedParameters,
    ExecutionResult,
} from "./types";

const STATE_DIR = path.join(process.env.APPDATA || process.env.HOME || ".", "AE-Knowledge-Vault", "learning-state");
const CASE_STORE_FILE = path.join(STATE_DIR, "case-store.json");
const DEFAULT_VALUE_STORE_FILE = path.join(STATE_DIR, "default-value-store.json");
const EXECUTION_RECORDS_FILE = path.join(STATE_DIR, "execution-records.json");
const CHECKPOINTS_DIR = path.join(STATE_DIR, "checkpoints");

interface PersistentStorage {
    load<T>(filePath: string, defaultValue: T): T;
    save<T>(filePath: string, data: T): void;
}

class FileSystemStorage implements PersistentStorage {
    load<T>(filePath: string, defaultValue: T): T {
        try {
            if (!fs.existsSync(filePath)) {
                return defaultValue;
            }
            const content = fs.readFileSync(filePath, "utf-8");
            return JSON.parse(content) as T;
        } catch {
            return defaultValue;
        }
    }

    save<T>(filePath: string, data: T): void {
        const dir = path.dirname(filePath);
        if (!fs.existsSync(dir)) {
            fs.mkdirSync(dir, { recursive: true });
        }
        fs.writeFileSync(filePath, JSON.stringify(data, null, 2), "utf-8");
    }
}

export interface CaseStore {
    addTemplate(template: ParameterTemplate): void;
    findTemplates(effectMatchName: string): ParameterTemplate[];
    incrementUsage(templateId: string): void;
    getAllTemplates(): ParameterTemplate[];
}

export class PersistentCaseStore implements CaseStore {
    private storage: PersistentStorage;
    private templates: Map<string, ParameterTemplate> = new Map();

    constructor(storage: PersistentStorage = new FileSystemStorage()) {
        this.storage = storage;
        this.load();
    }

    private load(): void {
        const saved = this.storage.load<ParameterTemplate[]>(CASE_STORE_FILE, []);
        for (const t of saved) {
            this.templates.set(t.id, t);
        }
    }

    private save(): void {
        this.storage.save(CASE_STORE_FILE, Array.from(this.templates.values()));
    }

    addTemplate(template: ParameterTemplate): void {
        this.templates.set(template.id, template);
        this.save();
    }

    findTemplates(effectMatchName: string): ParameterTemplate[] {
        return Array.from(this.templates.values())
            .filter((t) => t.effectMatchName === effectMatchName)
            .sort((a, b) => b.usageCount - a.usageCount);
    }

    incrementUsage(templateId: string): void {
        const t = this.templates.get(templateId);
        if (t) {
            t.usageCount++;
            t.lastUsed = new Date().toISOString();
            this.save();
        }
    }

    getAllTemplates(): ParameterTemplate[] {
        return Array.from(this.templates.values());
    }

    replaceTemplates(templates: ParameterTemplate[]): void {
        this.templates.clear();
        for (const t of templates) {
            this.templates.set(t.id, t);
        }
        this.save();
    }
}

export interface DefaultValueStore {
    get(effectMatchName: string, paramName: string): number | string | boolean | number[] | undefined;
    update(
        effectMatchName: string,
        paramName: string,
        actualValue: number | string | boolean | number[],
        learningRate: number,
    ): void;
    getAll(effectMatchName: string): Record<string, number | string | boolean | number[]>;
}

export class PersistentDefaultValueStore implements DefaultValueStore {
    private storage: PersistentStorage;
    private store: Map<string, { value: number | string | boolean | number[]; weight: number }> = new Map();

    constructor(storage: PersistentStorage = new FileSystemStorage()) {
        this.storage = storage;
        this.load();
    }

    private load(): void {
        const saved = this.storage.load<Record<string, { value: number | string | boolean | number[]; weight: number }>>(
            DEFAULT_VALUE_STORE_FILE,
            {}
        );
        for (const [key, value] of Object.entries(saved)) {
            this.store.set(key, value);
        }
    }

    private save(): void {
        const obj: Record<string, { value: number | string | boolean | number[]; weight: number }> = {};
        const entries = Array.from(this.store.entries());
        for (let i = 0; i < entries.length; i++) {
            obj[entries[i][0]] = entries[i][1];
        }
        this.storage.save(DEFAULT_VALUE_STORE_FILE, obj);
    }

    get(effectMatchName: string, paramName: string): number | string | boolean | number[] | undefined {
        return this.store.get(`${effectMatchName}.${paramName}`)?.value;
    }

    update(
        effectMatchName: string,
        paramName: string,
        actualValue: number | string | boolean | number[],
        learningRate: number,
    ): void {
        const key = `${effectMatchName}.${paramName}`;
        const existing = this.store.get(key);

        if (!existing) {
            this.store.set(key, { value: actualValue, weight: learningRate });
            this.save();
            return;
        }

        if (typeof existing.value === "number" && typeof actualValue === "number") {
            const oldWeight = existing.weight;
            const newWeight = learningRate;
            const totalWeight = oldWeight + newWeight;
            const newValue = (existing.value * oldWeight + actualValue * newWeight) / totalWeight;
            this.store.set(key, { value: newValue, weight: totalWeight });
        } else {
            this.store.set(key, { value: actualValue, weight: learningRate });
        }
        this.save();
    }

    getAll(effectMatchName: string): Record<string, number | string | boolean | number[]> {
        const result: Record<string, number | string | boolean | number[]> = {};
        const entries = Array.from(this.store.entries());
        for (let i = 0; i < entries.length; i++) {
            const key = entries[i][0];
            const entry = entries[i][1];
            if (key.startsWith(`${effectMatchName}.`)) {
                const paramName = key.slice(effectMatchName.length + 1);
                result[paramName] = entry.value;
            }
        }
        return result;
    }

    getAllEntries(): Map<string, { value: number | string | boolean | number[]; weight: number }> {
        return new Map(this.store);
    }

    replaceAll(entries: Record<string, { value: number | string | boolean | number[]; weight: number }>): void {
        this.store.clear();
        for (const [key, value] of Object.entries(entries)) {
            this.store.set(key, value);
        }
        this.save();
    }
}

export class PersistentLearningLoop {
    private caseStore: CaseStore;
    private defaultValueStore: DefaultValueStore;
    private confidenceAdjustments: ConfidenceAdjustment[] = [];
    private executionRecords: ExecutionRecord[] = [];
    private storage: PersistentStorage;

    private learningRate: number;
    private boostDelta: number;
    private penalizeDelta: number;

    constructor(
        caseStore?: CaseStore,
        defaultValueStore?: DefaultValueStore,
        options?: {
            learningRate?: number;
            boostDelta?: number;
            penalizeDelta?: number;
        },
        storage: PersistentStorage = new FileSystemStorage()
    ) {
        this.storage = storage;
        this.caseStore = caseStore || new PersistentCaseStore(storage);
        this.defaultValueStore = defaultValueStore || new PersistentDefaultValueStore(storage);
        this.learningRate = options?.learningRate ?? 0.3;
        this.boostDelta = options?.boostDelta ?? 0.05;
        this.penalizeDelta = options?.penalizeDelta ?? 0.1;

        this.loadExecutionRecords();
        this.loadConfidenceAdjustments();
    }

    private loadExecutionRecords(): void {
        this.executionRecords = this.storage.load<ExecutionRecord[]>(EXECUTION_RECORDS_FILE, []);
    }

    private saveExecutionRecords(): void {
        this.storage.save(EXECUTION_RECORDS_FILE, this.executionRecords);
    }

    private loadConfidenceAdjustments(): void {
        const records = this.storage.load<ExecutionRecord[]>(EXECUTION_RECORDS_FILE, []);
        for (const record of records) {
            if (record.reasoningPath && record.reasoningPath.length > 0) {
                const direction = record.userUndone || !record.execution.success ? "penalize" : "boost";
                this.confidenceAdjustments.push({
                    reasoningPath: record.reasoningPath,
                    direction,
                    delta: direction === "boost" ? this.boostDelta : this.penalizeDelta,
                    reason: record.execution.success
                        ? `执行成功 (record: ${record.id})`
                        : `执行失败 (record: ${record.id})`,
                    timestamp: record.timestamp,
                });
            }
        }
    }

    saveCheckpoint(name: string): string {
        if (!fs.existsSync(CHECKPOINTS_DIR)) {
            fs.mkdirSync(CHECKPOINTS_DIR, { recursive: true });
        }

        const defaultValuesObj: Record<string, { value: number | string | boolean | number[]; weight: number }> = {};
        const dvStore = this.defaultValueStore as PersistentDefaultValueStore;
        if (dvStore && "getAllEntries" in dvStore) {
            const entries = (dvStore as any).getAllEntries();
            for (const [key, value] of entries) {
                defaultValuesObj[key] = value;
            }
        }

        const checkpoint = {
            name,
            timestamp: new Date().toISOString(),
            executionRecords: this.executionRecords,
            confidenceAdjustments: this.confidenceAdjustments,
            templates: this.caseStore.getAllTemplates(),
            defaultValues: defaultValuesObj,
        };

        const filePath = path.join(CHECKPOINTS_DIR, `${name}_${Date.now()}.json`);
        fs.writeFileSync(filePath, JSON.stringify(checkpoint, null, 2), "utf-8");

        return filePath;
    }

    loadCheckpoint(filePath: string): boolean {
        try {
            const content = fs.readFileSync(filePath, "utf-8");
            const checkpoint = JSON.parse(content);

            if (checkpoint.executionRecords) {
                this.executionRecords = checkpoint.executionRecords;
            }
            if (checkpoint.confidenceAdjustments) {
                this.confidenceAdjustments = checkpoint.confidenceAdjustments;
            }
            if (checkpoint.templates && Array.isArray(checkpoint.templates)) {
                const cs = this.caseStore as PersistentCaseStore;
                if (cs && "replaceTemplates" in cs) {
                    (cs as any).replaceTemplates(checkpoint.templates);
                }
            }
            if (checkpoint.defaultValues && typeof checkpoint.defaultValues === "object") {
                const dvStore = this.defaultValueStore as PersistentDefaultValueStore;
                if (dvStore && "replaceAll" in dvStore) {
                    (dvStore as any).replaceAll(checkpoint.defaultValues);
                }
            }

            this.saveExecutionRecords();
            return true;
        } catch {
            return false;
        }
    }

    recordExecution(
        userInput: string,
        intentType: string,
        expected: ExpectedParameters,
        execution: ExecutionResult,
        verification: VerificationResult,
        userFeedback?: {
            satisfied?: boolean;
            adjusted?: boolean;
            undone?: boolean;
            finalParams?: Array<{ name: string; value: number | string | boolean | number[] }>;
        },
        reasoningPath?: string[],
    ): ExecutionRecord {
        const record: ExecutionRecord = {
            id: `exec_${Date.now()}_${Math.floor(Math.random() * 10000)}`,
            timestamp: new Date().toISOString(),
            userInput,
            intentType,
            expected,
            execution,
            verification,
            userSatisfied: userFeedback?.satisfied,
            userAdjusted: userFeedback?.adjusted,
            userUndone: userFeedback?.undone,
            finalParams: userFeedback?.finalParams,
            reasoningPath,
        };

        this.executionRecords.push(record);
        this.saveExecutionRecords();

        if (this.executionRecords.length > 1000) {
            this.executionRecords = this.executionRecords.slice(-500);
            this.saveExecutionRecords();
        }

        if (userFeedback?.undone) {
            this.learnFromFailure(record);
        } else if (userFeedback?.adjusted && userFeedback.finalParams) {
            this.learnFromDeviation(record, userFeedback.finalParams);
        } else if (userFeedback?.satisfied || (execution.success && verification.passed)) {
            this.learnFromSuccess(record);
        } else if (!execution.success) {
            this.learnFromFailure(record);
        }

        return record;
    }

    private learnFromSuccess(record: ExecutionRecord): void {
        const { expected, execution } = record;

        if (expected.effectMatchName && execution.effectName) {
            const template: ParameterTemplate = {
                id: `tpl_${Date.now()}_${Math.floor(Math.random() * 10000)}`,
                source: "auto-learned",
                effectMatchName: expected.effectMatchName,
                effectName: execution.effectName,
                parameters: this.propsToObject(expected.properties),
                userRating: "positive",
                usageCount: 1,
                lastUsed: new Date().toISOString(),
                sourceInput: record.userInput,
            };
            this.caseStore.addTemplate(template);
        }

        if (record.reasoningPath && record.reasoningPath.length > 0) {
            const adjustment: ConfidenceAdjustment = {
                reasoningPath: record.reasoningPath,
                direction: "boost",
                delta: this.boostDelta,
                reason: `执行成功 (record: ${record.id})`,
                timestamp: new Date().toISOString(),
            };
            this.confidenceAdjustments.push(adjustment);
        }
    }

    private learnFromDeviation(
        record: ExecutionRecord,
        finalParams: Array<{ name: string; value: number | string | boolean | number[] }>,
    ): void {
        const { expected } = record;
        if (!expected.effectMatchName) return;

        for (const finalParam of finalParams) {
            const expectedParam = expected.properties.find((p) => p.name === finalParam.name);
            if (!expectedParam) continue;

            const deviation = this.calculateParamDeviation(expectedParam.value, finalParam.value);

            if (deviation > 0.05) {
                this.defaultValueStore.update(
                    expected.effectMatchName,
                    finalParam.name,
                    finalParam.value,
                    this.learningRate,
                );
            }
        }
    }

    private learnFromFailure(record: ExecutionRecord): void {
        if (record.reasoningPath && record.reasoningPath.length > 0) {
            const adjustment: ConfidenceAdjustment = {
                reasoningPath: record.reasoningPath,
                direction: "penalize",
                delta: this.penalizeDelta,
                reason: record.execution.success
                    ? `用户撤销 (record: ${record.id})`
                    : `执行失败: ${record.execution.errorCode || "unknown"} (record: ${record.id})`,
                timestamp: new Date().toISOString(),
            };
            this.confidenceAdjustments.push(adjustment);
        }
    }

    private calculateParamDeviation(
        expected: number | string | boolean | number[],
        actual: number | string | boolean | number[],
    ): number {
        if (typeof expected === "number" && typeof actual === "number") {
            if (expected === 0) return Math.abs(actual);
            return Math.abs(expected - actual) / Math.abs(expected);
        }
        if (Array.isArray(expected) && Array.isArray(actual)) {
            if (expected.length === 0) return 0;
            const deviations = expected.map((e, i) => {
                const a = actual[i] as number;
                const eNum = e as number;
                if (eNum === 0) return Math.abs(a);
                return Math.abs(eNum - a) / Math.abs(eNum);
            });
            return deviations.reduce((sum, d) => sum + d, 0) / deviations.length;
        }
        return expected === actual ? 0 : 1;
    }

    private propsToObject(
        props: Array<{ name: string; value: number | string | boolean | number[] }>,
    ): Record<string, number | string | boolean | number[]> {
        const result: Record<string, number | string | boolean | number[]> = {};
        for (const p of props) {
            result[p.name] = p.value;
        }
        return result;
    }

    getMetrics(): LearningMetrics {
        const total = this.executionRecords.length;
        const successCount = this.executionRecords.filter(
            (r) => r.execution.success && r.verification.passed,
        ).length;
        const failureCount = this.executionRecords.filter((r) => !r.execution.success).length;
        const userAdjustedCount = this.executionRecords.filter((r) => r.userAdjusted).length;
        const userUndoneCount = this.executionRecords.filter((r) => r.userUndone).length;

        const deviationSum = this.executionRecords
            .filter((r) => r.verification.deviationScore !== undefined)
            .reduce((sum, r) => sum + r.verification.deviationScore, 0);
        const deviationCount = this.executionRecords.filter(
            (r) => r.verification.deviationScore !== undefined,
        ).length;
        const averageDeviation = deviationCount > 0 ? deviationSum / deviationCount : 0;

        const learnedTemplates = this.caseStore.getAllTemplates().length;
        const confidenceAdjustments = this.confidenceAdjustments.length;

        const accuracyImprovement = this.calculateAccuracyImprovement();

        return {
            totalExecutions: total,
            successCount,
            failureCount,
            userAdjustedCount,
            userUndoneCount,
            successRate: total > 0 ? successCount / total : 0,
            averageDeviation,
            learnedTemplates,
            confidenceAdjustments,
            accuracyImprovement,
        };
    }

    private calculateAccuracyImprovement(): number {
        const records = this.executionRecords;
        if (records.length < 4) return 0;

        const recentCount = Math.min(10, Math.floor(records.length / 2));
        const recentRecords = records.slice(-recentCount);
        const earlierRecords = records.slice(-recentCount * 2, -recentCount);

        if (earlierRecords.length === 0) return 0;

        const recentSuccess = recentRecords.filter((r) => r.execution.success).length / recentRecords.length;
        const earlierSuccess = earlierRecords.filter((r) => r.execution.success).length / earlierRecords.length;

        return recentSuccess - earlierSuccess;
    }

    getExecutionRecords(): ExecutionRecord[] {
        return [...this.executionRecords];
    }

    getConfidenceAdjustments(): ConfidenceAdjustment[] {
        return [...this.confidenceAdjustments];
    }

    getCaseStore(): CaseStore {
        return this.caseStore;
    }

    getDefaultValueStore(): DefaultValueStore {
        return this.defaultValueStore;
    }

    getStateSummary(): {
        totalExecutions: number;
        templatesCount: number;
        lastSaved: string;
        checkpointsCount: number;
    } {
        let checkpointsCount = 0;
        try {
            if (fs.existsSync(CHECKPOINTS_DIR)) {
                checkpointsCount = fs.readdirSync(CHECKPOINTS_DIR).length;
            }
        } catch {
            checkpointsCount = 0;
        }

        return {
            totalExecutions: this.executionRecords.length,
            templatesCount: this.caseStore.getAllTemplates().length,
            lastSaved: new Date().toISOString(),
            checkpointsCount,
        };
    }
}

export const persistentLearningLoop = new PersistentLearningLoop();

File: learning/persistent_learning_loop.py
"""
Persistent Learning Loop
Phase 5 - 持久化学习循环

在 learning_loop.py 的基础上增加：
  1. JSON文件持久化存储
  2. 自动保存/加载机制（每次操作后自动保存）
  3. 检查点机制（定期保存检查点）
  4. 跨IDE共享（文件系统存储）

对应架构设计文档 7.5 节 LearningLoop + 持久化扩展
"""

import os
import json
import time
import random
import threading
from datetime import datetime
from typing import Dict, List, Optional, Any, Tuple
from dataclasses import dataclass, field, asdict

from learning_loop import (
    LearningLoop,
    CaseStore,
    DefaultValueStore,
    ParameterTemplate,
    ExpectedParameters,
    ExecutionResult as LRExecutionResult,
    VerificationResult as LRVerificationResult,
    UserFeedback,
    ExecutionRecord,
    ConfidenceAdjustment,
)


def _get_state_dir() -> str:
    """获取持久化状态目录（跨平台）"""
    if os.name == "nt":
        base = os.environ.get("APPDATA", os.path.expanduser("~"))
    else:
        base = os.environ.get("HOME", os.path.expanduser("~"))
    return os.path.join(base, "AE-Knowledge-Vault", "learning-state")


STATE_DIR = _get_state_dir()
CASE_STORE_FILE = os.path.join(STATE_DIR, "case-store.json")
DEFAULT_VALUE_STORE_FILE = os.path.join(STATE_DIR, "default-value-store.json")
EXECUTION_RECORDS_FILE = os.path.join(STATE_DIR, "execution-records.json")
CONFIDENCE_ADJUSTMENTS_FILE = os.path.join(STATE_DIR, "confidence-adjustments.json")
CHECKPOINTS_DIR = os.path.join(STATE_DIR, "checkpoints")

MAX_RECORDS = 10000
CHECKPOINT_INTERVAL_SEC = 3600


class FileSystemStorage:
    """文件系统持久化存储"""

    def load(self, file_path: str, default_value: Any) -> Any:
        try:
            if not os.path.exists(file_path):
                return default_value
            with open(file_path, "r", encoding="utf-8") as f:
                return json.load(f)
        except Exception:
            return default_value

    def save(self, file_path: str, data: Any) -> None:
        dir_path = os.path.dirname(file_path)
        if not os.path.exists(dir_path):
            os.makedirs(dir_path, exist_ok=True)
        with open(file_path, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)


class PersistentCaseStore(CaseStore):
    """持久化参数模板存储"""

    def __init__(self, storage: FileSystemStorage):
        super().__init__()
        self._storage = storage
        self._templates: Dict[str, ParameterTemplate] = {}
        self._load()

    def _load(self) -> None:
        saved = self._storage.load(CASE_STORE_FILE, [])
        for t_data in saved:
            try:
                template = ParameterTemplate(
                    id=t_data.get("id", f"tpl_{int(time.time()*1000)}_{random.randint(0,9999)}"),
                    source=t_data.get("source", "loaded"),
                    effect_match_name=t_data.get("effectMatchName", ""),
                    effect_name=t_data.get("effectName", ""),
                    parameters=t_data.get("parameters", {}),
                    user_rating=t_data.get("userRating", "neutral"),
                    usage_count=t_data.get("usageCount", 0),
                    last_used=t_data.get("lastUsed", ""),
                    source_input=t_data.get("sourceInput", ""),
                )
                self._templates[template.id] = template
            except Exception:
                continue

    def _save(self) -> None:
        data = []
        for t in self._templates.values():
            data.append({
                "id": t.id,
                "source": t.source,
                "effectMatchName": t.effect_match_name,
                "effectName": t.effect_name,
                "parameters": t.parameters,
                "userRating": t.user_rating,
                "usageCount": t.usage_count,
                "lastUsed": t.last_used,
                "sourceInput": t.source_input,
            })
        self._storage.save(CASE_STORE_FILE, data)

    def add_template(self, template: ParameterTemplate) -> None:
        self._templates[template.id] = template
        self._save()

    def find_templates(self, effect_match_name: str) -> List[ParameterTemplate]:
        return sorted(
            [t for t in self._templates.values()
             if t.effect_match_name == effect_match_name],
            key=lambda t: t.usage_count,
            reverse=True,
        )

    def increment_usage(self, template_id: str) -> None:
        t = self._templates.get(template_id)
        if t:
            t.usage_count += 1
            t.last_used = datetime.now().isoformat()
            self._save()

    def get_all_templates(self) -> List[ParameterTemplate]:
        return list(self._templates.values())


class PersistentDefaultValueStore(DefaultValueStore):
    """持久化默认值存储"""

    def __init__(self, storage: FileSystemStorage):
        super().__init__()
        self._storage = storage
        self._store: Dict[str, Tuple[Any, float]] = {}
        self._load()

    def _load(self) -> None:
        saved = self._storage.load(DEFAULT_VALUE_STORE_FILE, {})
        for key, entry in saved.items():
            if isinstance(entry, dict):
                self._store[key] = (entry.get("value"), entry.get("weight", 1.0))
            elif isinstance(entry, list) and len(entry) >= 2:
                self._store[key] = (entry[0], entry[1])

    def _save(self) -> None:
        data = {}
        for key, (value, weight) in self._store.items():
            data[key] = {"value": value, "weight": weight}
        self._storage.save(DEFAULT_VALUE_STORE_FILE, data)

    def get(self, effect_match_name: str, param_name: str) -> Any:
        key = f"{effect_match_name}.{param_name}"
        entry = self._store.get(key)
        return entry[0] if entry else None

    def update(self, effect_match_name: str, param_name: str, value: Any, weight: float = 1.0) -> None:
        key = f"{effect_match_name}.{param_name}"
        self._store[key] = (value, weight)
        self._save()

    def get_all(self) -> Dict[str, Any]:
        result = {}
        for key, (value, _) in self._store.items():
            result[key] = value
        return result


class PersistentLearningLoop(LearningLoop):
    """持久化学习循环 - 继承 LearningLoop 并添加持久化能力"""

    def __init__(self, storage: Optional[FileSystemStorage] = None,
                 auto_save: bool = True,
                 checkpoint_enabled: bool = True):
        self._pstorage = storage or FileSystemStorage()
        self._auto_save = auto_save
        self._checkpoint_enabled = checkpoint_enabled
        self._last_checkpoint_time = time.time()
        self._lock = threading.Lock()

        case_store = PersistentCaseStore(self._pstorage)
        default_value_store = PersistentDefaultValueStore(self._pstorage)

        super().__init__(case_store=case_store, default_value_store=default_value_store)

        self._load_execution_records()
        self._load_confidence_adjustments()

    # ========== 执行记录持久化 ==========

    def _load_execution_records(self) -> None:
        saved = self._pstorage.load(EXECUTION_RECORDS_FILE, [])
        self._execution_records = []
        for r_data in saved:
            try:
                record = self._dict_to_record(r_data)
                self._execution_records.append(record)
            except Exception:
                continue

    def _save_execution_records(self) -> None:
        records_to_save = self._execution_records[-MAX_RECORDS:]
        data = [self._record_to_dict(r) for r in records_to_save]
        self._pstorage.save(EXECUTION_RECORDS_FILE, data)

    def _load_confidence_adjustments(self) -> None:
        saved = self._pstorage.load(CONFIDENCE_ADJUSTMENTS_FILE, [])
        self._confidence_adjustments = []
        for a_data in saved:
            try:
                self._confidence_adjustments.append(ConfidenceAdjustment(
                    reasoning_path=a_data.get("reasoningPath", []),
                    direction=a_data.get("direction", "boost"),
                    delta=a_data.get("delta", 0.05),
                    timestamp=a_data.get("timestamp", ""),
                ))
            except Exception:
                continue

    def _save_confidence_adjustments(self) -> None:
        data = []
        for a in self._confidence_adjustments:
            data.append({
                "reasoningPath": a.reasoning_path,
                "direction": a.direction,
                "delta": a.delta,
                "timestamp": a.timestamp,
            })
        self._pstorage.save(CONFIDENCE_ADJUSTMENTS_FILE, data)

    def _dict_to_record(self, data: Dict[str, Any]) -> ExecutionRecord:
        expected = ExpectedParameters(
            comp_name=data.get("expected", {}).get("compName", ""),
            layer_index=data.get("expected", {}).get("layerIndex", 0),
            effect_match_name=data.get("expected", {}).get("effectMatchName", ""),
            effect_name=data.get("expected", {}).get("effectName", ""),
            properties=[],
        )
        execution = LRExecutionResult(
            success=data.get("execution", {}).get("success", False),
            error_code=data.get("execution", {}).get("errorCode", ""),
            error_message=data.get("execution", {}).get("errorMessage", ""),
            effect_name=data.get("execution", {}).get("effectName", ""),
            execution_time=data.get("execution", {}).get("executionTime", 0),
        )
        verification = LRVerificationResult(
            passed=data.get("verification", {}).get("passed", False),
            deviation_score=data.get("verification", {}).get("deviationScore"),
            mismatched_properties=data.get("verification", {}).get("mismatchedProperties", []),
            verified_at=data.get("verification", {}).get("verifiedAt", ""),
        )
        return ExecutionRecord(
            id=data.get("id", ""),
            timestamp=data.get("timestamp", ""),
            user_input=data.get("userInput", ""),
            intent_type=data.get("intentType", ""),
            expected=expected,
            execution=execution,
            verification=verification,
            user_satisfied=data.get("userSatisfied"),
            user_adjusted=data.get("userAdjusted"),
            user_undone=data.get("userUndone"),
            final_params=data.get("finalParams"),
            reasoning_path=data.get("reasoningPath"),
        )

    def _record_to_dict(self, record: ExecutionRecord) -> Dict[str, Any]:
        expected = record.expected
        expected_dict = {
            "compName": getattr(expected, "comp_name", ""),
            "layerIndex": getattr(expected, "layer_index", 0),
            "effectMatchName": getattr(expected, "effect_match_name", ""),
            "effectName": getattr(expected, "effect_name", ""),
            "properties": [
                {"name": p.name, "value": p.value, "tolerance": getattr(p, "tolerance", 0.01)}
                for p in getattr(expected, "properties", [])
            ],
        }

        execution = record.execution
        execution_dict = {
            "success": getattr(execution, "success", False),
            "errorCode": getattr(execution, "error_code", ""),
            "errorMessage": getattr(execution, "error_message", ""),
            "effectName": getattr(execution, "effect_name", ""),
            "executionTime": getattr(execution, "execution_time", 0),
        }

        verification = record.verification
        verification_dict = {
            "passed": getattr(verification, "passed", False),
            "deviationScore": getattr(verification, "deviation_score"),
            "mismatchedProperties": getattr(verification, "mismatched_properties", []),
            "verifiedAt": getattr(verification, "verified_at", ""),
        }

        return {
            "id": record.id,
            "userInput": record.user_input,
            "intentType": record.intent_type,
            "timestamp": record.timestamp,
            "expected": expected_dict,
            "execution": execution_dict,
            "verification": verification_dict,
            "userSatisfied": record.user_satisfied,
            "userAdjusted": record.user_adjusted,
            "userUndone": record.user_undone,
            "finalParams": record.final_params,
            "reasoningPath": record.reasoning_path,
        }

    # ========== 检查点机制 ==========

    def _maybe_save_checkpoint(self) -> None:
        if not self._checkpoint_enabled:
            return
        now = time.time()
        if now - self._last_checkpoint_time < CHECKPOINT_INTERVAL_SEC:
            return

        try:
            if not os.path.exists(CHECKPOINTS_DIR):
                os.makedirs(CHECKPOINTS_DIR, exist_ok=True)
            ts = datetime.now().strftime("%Y%m%d_%H%M%S")
            cp_dir = os.path.join(CHECKPOINTS_DIR, f"checkpoint_{ts}")
            os.makedirs(cp_dir, exist_ok=True)

            import shutil
            for fname in [
                "case-store.json", "default-value-store.json",
                "execution-records.json", "confidence-adjustments.json",
            ]:
                src = os.path.join(STATE_DIR, fname)
                if os.path.exists(src):
                    shutil.copy2(src, os.path.join(cp_dir, fname))

            checkpoints = sorted([
                d for d in os.listdir(CHECKPOINTS_DIR)
                if d.startswith("checkpoint_")
            ])
            for old_cp in checkpoints[:-10]:
                shutil.rmtree(os.path.join(CHECKPOINTS_DIR, old_cp), ignore_errors=True)

            self._last_checkpoint_time = now
        except Exception:
            pass

    def _trigger_save(self) -> None:
        if not self._auto_save:
            return
        with self._lock:
            try:
                self._save_execution_records()
                self._save_confidence_adjustments()
            except Exception:
                pass
            self._maybe_save_checkpoint()

    # ========== 覆盖父类方法 ==========

    def record_execution(
        self,
        user_input: str,
        intent_type: str,
        expected: ExpectedParameters,
        execution: LRExecutionResult,
        verification: LRVerificationResult,
        user_feedback: Optional[UserFeedback] = None,
        reasoning_path: Optional[List[str]] = None,
    ) -> ExecutionRecord:
        record = super().record_execution(
            user_input, intent_type, expected,
            execution, verification, user_feedback, reasoning_path
        )
        self._trigger_save()
        return record

    # ========== 额外方法 ==========

    def force_save(self) -> None:
        """强制立即保存所有状态"""
        with self._lock:
            self._save_execution_records()
            self._save_confidence_adjustments()
            if isinstance(self._case_store, PersistentCaseStore):
                self._case_store._save()
            if isinstance(self._default_value_store, PersistentDefaultValueStore):
                self._default_value_store._save()

    def load_from_checkpoint(self, checkpoint_name: str) -> bool:
        """从指定检查点恢复"""
        try:
            import shutil
            cp_dir = os.path.join(CHECKPOINTS_DIR, checkpoint_name)
            if not os.path.exists(cp_dir):
                return False
            for fname in [
                "case-store.json", "default-value-store.json",
                "execution-records.json", "confidence-adjustments.json",
            ]:
                src = os.path.join(cp_dir, fname)
                dst = os.path.join(STATE_DIR, fname)
                if os.path.exists(src):
                    shutil.copy2(src, dst)

            if isinstance(self._case_store, PersistentCaseStore):
                self._case_store._load()
            if isinstance(self._default_value_store, PersistentDefaultValueStore):
                self._default_value_store._load()
            self._load_execution_records()
            self._load_confidence_adjustments()
            return True
        except Exception:
            return False

    def list_checkpoints(self) -> List[str]:
        """列出所有检查点"""
        if not os.path.exists(CHECKPOINTS_DIR):
            return []
        return sorted([
            d for d in os.listdir(CHECKPOINTS_DIR)
            if d.startswith("checkpoint_")
        ], reverse=True)

    def get_stats(self) -> Dict[str, Any]:
        """获取持久化统计信息"""
        from learning_loop import LearningMetrics

        metrics: Dict[str, Any] = {
            "persistent": True,
            "state_dir": STATE_DIR,
            "auto_save": self._auto_save,
            "checkpoint_enabled": self._checkpoint_enabled,
            "checkpoints": len(self.list_checkpoints()),
            "templates_count": len(self._case_store.get_all_templates()),
            "execution_records_count": len(self._execution_records),
            "confidence_adjustments_count": len(self._confidence_adjustments),
        }
        return metrics
